reverse_replication/launcher.go

```diff
@@ -0,0 +1,273 @@
+	}
+	return nil
+}
+
+func main() {
+	ORDERING_TEMPLATE := "gs://deepchowdhury-gsql/templates/flex/Spanner_Change_Streams_to_Sink"
+	WRITER_TEMPLATE := "gs://deepchowdhury-gsql/templates/flex/Ordered_Changestream_Buffer_to_Sourcedb"
+	//WRITER_TEMPLATE := "gs://aks-test-revrep/images/ordered-changestream-buffer-to-sourcedb-image-spec.json"
+	setupGlobalFlags()
+	flag.Parse()
+
+	err := prechecks()
+	if err != nil {
+		fmt.Println("incorrect arguments passed:", err)
+		return
+	}
+	ctx := context.Background()
+	adminClient, _ := database.NewDatabaseAdminClient(ctx)
+	op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
+		Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", projectId, instanceId, dbName),
+		Statements: []string{fmt.Sprintf("CREATE CHANGE STREAM %s FOR ALL OPTIONS (value_capture_type = 'NEW_ROW')", changeStreamName)},
+	})
+	if err != nil {
+		fmt.Printf("Cannot submit request create change stream request: %v\n", err)
+		return
+	}
+	if err := op.Wait(ctx); err != nil {
+		if !strings.Contains(err.Error(), "Duplicate name in schema") {
+			fmt.Printf("Could not update database ddl: %v\n", err)
+			return
+		} else {
+			fmt.Printf("changestream with name '%s' already exists...skipping changestream creation\n", changeStreamName)
+		}
+	} else {
+		fmt.Println("Created changestream ", changeStreamName)
+	}
+
+	gcsclient, _ := storage.NewClient(ctx)
+	u, _ := url.Parse(sourceShardsFilePath)
+	rc, _ := gcsclient.Bucket(u.Host).Object(u.Path[1:]).NewReader(ctx)
+	bArr, _ := ioutil.ReadAll(rc)
+	rc.Close()
+	var data []interface{}
+	json.Unmarshal(bArr, &data)
+	arr := []string{}
+	for i := 0; i < len(data); i++ {
+		arr = append(arr, data[i].(map[string]interface{})["logicalShardId"].(string))
+	}
+
+	topicName := strings.Split(pubSubDataTopicId, "/")[3]
+	client, err := pubsub.NewClient(ctx, strings.Split(pubSubDataTopicId, "/")[1])
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer client.Close()
+	_, err = client.CreateTopic(ctx, topicName)
+	if err != nil {
+		if !(strings.Contains(err.Error(), "AlreadyExists")) {
+			fmt.Printf("could not create topic: %v\n", err)
+			return
+		} else {
+			fmt.Printf("topic '%s' already exists, skipping creation...\n", topicName)
+		}
+	} else {
+		fmt.Println("Created topic ", pubSubDataTopicId)
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < len(arr); i++ {
+		wg.Add(1)
+		go func(shardId string) {
+			defer wg.Done()
+			_, err := client.CreateSubscription(ctx, shardId, pubsub.SubscriptionConfig{
+				Topic:                 client.Topic(topicName),
+				AckDeadline:           600 * time.Second,
+				EnableMessageOrdering: true,
+				Filter:                fmt.Sprintf("attributes.shardId=\"%s\"", shardId),
+			})
+			if err != nil {
+				if !(strings.Contains(err.Error(), "AlreadyExists")) {
+					fmt.Printf("could not create subscription: %v\n", err)
+				} else {
+					fmt.Printf("subscription '%s' already exists, skipping creation...\n", shardId)
```

Comment:
Is there a way to check that the existing subscription is ordered and has the expected filter criteria?

Reply:
Same as above

Reply:
Reverse replication will fail incase the subscriber is not ordered and does not have the shard filter. If we are checking the existence of the subscriber in this script, then it should also check the properties of the same. It can be a TODO, but I believe the scope should be this script.